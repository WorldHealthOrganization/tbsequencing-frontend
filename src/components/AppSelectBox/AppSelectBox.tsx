/** @jsxImportSource @emotion/react */
import React, { useState } from 'react';
import { Popover } from '@mui/material';
import { SerializedStyles } from '@emotion/react';
import { AppButton, IButtonProps } from '../AppButton/AppButton';

export interface IDropDownContent {
  closeDropDown: () => void;
}

interface Props extends Pick<IButtonProps, 'children' | 'endIconName' | 'size' | 'variant' | 'style'> {
  dropdownContentRender: (e: IDropDownContent) => React.ReactNode
  popoverStyle?: SerializedStyles;
}

export const AppSelectBox = ({
  dropdownContentRender, children, endIconName, size, variant, style, popoverStyle,
}: Props) => {
  const [anchorEl, setAnchorEl] = useState<HTMLButtonElement | null>(null);

  const openDropDown = (e: React.MouseEvent<HTMLButtonElement>) => {
    setAnchorEl(e.currentTarget);
  };

  const closeDropDown = () => {
    setAnchorEl(null);
  };

  return (
    <>
      <AppButton
        onClick={openDropDown}
        style={style}
        variant={variant}
        size={size}
        endIconName={endIconName}
      >
        {children}
      </AppButton>
      <Popover
        css={popoverStyle}
        open={!!anchorEl}
        onClose={closeDropDown}
        anchorEl={anchorEl}
        anchorOrigin={{
          vertical: 'bottom',
          horizontal: 'center',
        }}
      >
        {dropdownContentRender({
          closeDropDown,
        })}
      </Popover>
    </>
  );
};

export default AppSelectBox;
