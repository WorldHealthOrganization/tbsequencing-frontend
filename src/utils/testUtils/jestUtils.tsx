import React, { ReactElement } from 'react';
import { render as rtlRender } from '@testing-library/react';
import { configureStore } from '@reduxjs/toolkit';
import { Provider } from 'react-redux';

import renderer from 'react-test-renderer';
import { BrowserRouter } from 'react-router-dom';
import { reducer } from '../../app/store';

const getWrapper = (
  ui: ReactElement,
  {
    // @ts-ignore
    preloadedState,
    store = configureStore({ reducer, preloadedState }),
  } = {},
) => ({ children }: { children: ReactElement }) => (
  <Provider store={store}>{children}</Provider>
);

function render(
  ui: ReactElement,
  {
    // @ts-ignore
    preloadedState,
    store = configureStore({ reducer, preloadedState }),
    ...renderOptions
  } = {},
) {
  const Wrapper = getWrapper(ui, { preloadedState, store });

  return rtlRender(ui, { wrapper: Wrapper, ...renderOptions });
}

const getTreeJSONWithRedux = (ui: ReactElement, {
  // @ts-ignore
  preloadedState,
  store = configureStore({ reducer, preloadedState }),
} = {}) => {
  const Wrapper = getWrapper(ui, { preloadedState, store });

  const tree = renderer
    .create(<Wrapper>{ui}</Wrapper>)
    .toJSON();

  return tree;
};

export const wrappedInRouter = (
  component: ReactElement,
) => <BrowserRouter>{component}</BrowserRouter>;

export * from '@testing-library/react';
export { render, getTreeJSONWithRedux };
