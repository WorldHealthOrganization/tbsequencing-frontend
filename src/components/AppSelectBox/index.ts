import AppSelectBox from './AppSelectBox';

export default AppSelectBox;
