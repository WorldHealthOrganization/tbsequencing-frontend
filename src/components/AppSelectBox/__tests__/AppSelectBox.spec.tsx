import React from 'react';
import { fireEvent, screen, waitFor } from '@testing-library/react';
import { getTreeJSONWithRedux, render } from '../../../utils/testUtils/jestUtils';
import AppSelectBox from '../AppSelectBox';

describe('AppSelectBox component test', () => {
  test('renders AppSelectBox component without error', () => {
    render(
      <AppSelectBox dropdownContentRender={() => <div>Content</div>} size="medium" variant="outlined">
        Select
      </AppSelectBox>,
    );

    const checkIfDropDownOpened = screen.queryByText('Content');
    expect(checkIfDropDownOpened).not.toBeInTheDocument();
  });

  test('renders dropdown content after click', () => {
    render(
      <AppSelectBox dropdownContentRender={() => <button type="button">Click</button>} size="medium" variant="outlined">
        Select
      </AppSelectBox>,
    );

    const selectBox = screen.queryByText('Select');

    if (selectBox) {
      fireEvent.click(selectBox);
    }

    const dropdownContent = screen.queryByText('Click');

    expect(selectBox).toBeInTheDocument();
    expect(dropdownContent).toBeInTheDocument();
  });

  test('passes closeDropDown func and closes dropdown', async () => {
    const mockOnClick = jest.fn();
    render(
      <AppSelectBox
        dropdownContentRender={({ closeDropDown }) => (
          <button
            onClick={() => {
              closeDropDown();
              mockOnClick();
            }}
            type="button"
          >
            Click
          </button>
        )}
        size="medium"
        variant="outlined"
      >
        Select
      </AppSelectBox>,
    );

    const selectBox = screen.queryByText('Select');

    if (selectBox) {
      fireEvent.click(selectBox);
    }

    const dropdownContent = screen.queryByText('Click');

    if (dropdownContent) {
      fireEvent.click(dropdownContent);
    }

    expect(mockOnClick).toHaveBeenCalled();
    expect(selectBox).toBeInTheDocument();
    await waitFor(() => {
      expect(dropdownContent).not.toBeInTheDocument();
    });
  });

  it('Should render correctly', () => {
    const tree = getTreeJSONWithRedux(
      <AppSelectBox dropdownContentRender={() => <div>Content</div>} size="medium" variant="outlined">
        Select
      </AppSelectBox>,
    );

    expect(tree).toMatchSnapshot();
  });
});
