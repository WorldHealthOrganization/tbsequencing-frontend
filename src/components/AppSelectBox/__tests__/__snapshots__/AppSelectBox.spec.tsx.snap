// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AppSelectBox component test Should render correctly 1`] = `
<button
  className="MuiButton-root MuiButton-outlined MuiButton-outlinedPrimary MuiButton-sizeMedium MuiButton-outlinedSizeMedium MuiButton-disableElevation MuiButtonBase-root css-sfprtt-MuiButtonBase-root-MuiButton-root-AppButton"
  disabled={false}
  onBlur={[Function]}
  onClick={[Function]}
  onContextMenu={[Function]}
  onDragLeave={[Function]}
  onFocus={[Function]}
  onKeyDown={[Function]}
  onKeyUp={[Function]}
  onMouseDown={[Function]}
  onMouseLeave={[Function]}
  onMouseUp={[Function]}
  onTouchEnd={[Function]}
  onTouchMove={[Function]}
  onTouchStart={[Function]}
  tabIndex={0}
  type="button"
>
  <p
    className="MuiTypography-root MuiTypography-body1 css-3cwtjx-MuiTypography-root-BaseText"
  >
    Select
  </p>
</button>
`;
